Coding style updates for PSR-12 compliance & improved readability

BudsieBuds opened the pull request:
This PR started as a minor update to the `.editorconfig` but evolved into a refactor to enhance the adherence to PSR-12 standards and more style consistency, following #4196. Including:
- An updated .editorconfig to enforce consistent styling across the codebase
- Adjustments to spacing and other formatting elements to align with best practices
- Overall enhancements improving code readability and implementing modern standards

While I've carefully checked everything, I could use an extra set of eyes to read over the changes. These updates should be safe to merge though.
I believe the timing is ideal, given the extensive changes in #4196, which will likely require PRs to be rebased anyway.

objecttothis commented:
@BudsieBuds I didn't see this PR until just now.  I just created one changing the .editorconfig.  I'll review this PR and we can go from there.

objecttothis commented:
closing #3708

objecttothis reviewed (changes requested):
OK, I had time to get through 27 of the files.  I'll do more probably Monday, but look at the comments I made because IMO, we need to make a customization to the linter
- Enforce each brace `{` `}` getting it's own line always (see my comments about it.

I feel less strong about these but we need to be consistent across the board
- No space between constructs (if, foreach, elseif, etc) and the opening parenthesis `(`
- No space between // and the comment.

objecttothis commented:
This PR doesn't cover it, and I think it's something we talked about doing as people make code changes, but PSR-12 also dictates that function names be camelCase not snake_case.

BudsieBuds commented on .editorconfig:
```diff
@@ -1,22 +1,15 @@
-indent_style = tab
\ No newline at end of file
+[*.md]
+trim_trailing_whitespace = false
```
In Markdown, two or more spaces at the end of a line create a line break when rendered. Removing trailing whitespace would alter the formatting.

This is less of a problem, but there's also the fact that it would remove the trailing whitespace from a codeblock. So this code has a space at the end. This could be needed.
```php
<?= "Hello, world!"; ?> 
```
By preserving trailing whitespace in Markdown files, we ensure that both formatting and code examples remain accurate and true to their original intent.

objecttothis replied:
OK, let's leave it in then.

BudsieBuds commented on .env-example:
```diff
@@ -59,5 +59,5 @@ app.db_log_enabled = false
 honeypot.label = 'Fill This Field'
 honeypot.name = 'honeypot'
-honeypot.template = '<label>{label}</label><input type="text" name="{name}" value=""/>'
+honeypot.template = '<label>{label}</label><input type="text" name="{name}" value="">'
```
Self-closing tags do not require a trailing slash in HTML. Only XML, XHTML, and SVG require self-closing tags. More info here: https://developer.mozilla.org/en-US/docs/Glossary/Void_element

objecttothis replied:
> Self-closing tags do not require a trailing slash in HTML. Only XML, XHTML, and SVG require self-closing tags. More info here: https://developer.mozilla.org/en-US/docs/Glossary/Void_element

I learned something new.  This stack exchange question also backs this up: https://stackoverflow.com/questions/3558119/are-non-void-self-closing-tags-valid-in-html5.  Here are two thoughts to consider:
- For readability the trailing slash on the void tag `<input />` does usually cause someone reading the code not to look down any further for a closing tag to see what is between the opening and closing tags.  Input isn't a great example because it doesn't have closing tags (it's a void tag) but anyone not registering that as they are skimming code may still look, therefore slowing down their search of the code.
- This code block actually came from CodeIgniter that way (https://codeigniter.com/user_guide/libraries/honeypot.html#customizing-honeypot). That said, they must have removed the trailing slash at some point from the code base.  Another reason we ought to follow suit.

BudsieBuds commented on app/Config/Database.php:
```diff
@@ -122,8 +122,7 @@ public function __construct()
         // we don't overwrite live data on accident.
-        switch(ENVIRONMENT)
-        {
+        switch (ENVIRONMENT) {
```
Based on my understanding of PSR-12 (which is new), the standard specifies that braces [for control structures go on the same line](https://www.php-fig.org/psr/psr-12/#5-control-structures) and [for methods and functions go on a new line](https://www.php-fig.org/psr/psr-12/#44-methods-and-functions).

objecttothis replied:
> Based on my understanding of PSR-12 (which is new), the standard specifies that braces [for control structures go on the same line](https://www.php-fig.org/psr/psr-12/#5-control-structures) and [for methods and functions go on a new line](https://www.php-fig.org/psr/psr-12/#44-methods-and-functions).

Yes, you are correct in your reading of the PSR-12 standard.  It's really the only thing in the the standard that I think looks horribly ugly and reduces readability. It's almost like they said "hey let's make the code wider horizontally but shorter vertically because 'reasons'".  Take this section of code from the expenses controller for example:
```
            if (!empty($item_kit_items)) {
                $success = $this->item_kit_items->save_value($item_kit_items, $item_kit_id);
            } else {
                $success = true;
            }

            if ($new_item) {
                echo json_encode([
                    'success' => $success,
                    'message' => lang('Item_kits.successful_adding') . ' ' . $item_kit_data['name'],
                    'id' => $item_kit_id
                ]);
            } else {
                echo json_encode([
                    'success' => $success,
                    'message' => lang('Item_kits.successful_updating') . ' ' . $item_kit_data['name'],
                    'id' => $item_kit_id
                ]);
            }
```

And now the original code
```
           if (!empty($item_kit_items))
            {
                $success = $this->item_kit_items->save_value($item_kit_items, $item_kit_id);
            }
            else
            {
                $success = true;
            }

            if($new_item)
            {
                echo json_encode ([
                    'success' => $success,
                    'message' => lang('Item_kits.successful_adding').' '.$item_kit_data['name'],
                    'id' => $item_kit_id
                ]);
            }
            else
            {
                echo json_encode ([
                    'success' => $success,
                    'message' => lang('Item_kits.successful_updating').' '.$item_kit_data['name'],
                    'id' => $item_kit_id
                ]);
            }
```

It's difficult for me to see how the PSR-12 version is not more difficult for people to follow what is inside the braces of which construct.

In the end, PSR-12 is a coding standard and there is a lot of good in it, but it's just a standard.  It is indeed possible to customize the linter to allow variations.  While I won't revolt if @jekkos and everyone says "no we need to rigidly follow PSR-12, but this is the only thing that I REALLY don't like about PSR-12 and my vote, if I have one, is to customize the linter to instead enforce the convention we have been using in the code which is every construct and function gets it's own line for braces.  All other objections I had to the PSR-12 standard, I will acquiesce to.

jekkos replied:
Im also a bit in doubt here. If we do this exception then we should have an automatic check that fails on such formatting so that it is clear for contributors what the rules are. If we can accomplish that then fine, I don't have strong feelings to prefer one over the other. If it does not ask too much extra work then its fine.

objecttothis replied:
The only thing I don't like about making this exception is the extra work it will take in customizing the linter and in enforcing the exception in the future. It will result in more rejected PRs than a straightforward "make it PSR-12 compliant." That and EVERYONE will need to manage the same exceptions in their IDEs.  Let's follow strict PSR-12 and I'll get over it.

objecttothis replied:
> I noticed that the CI adheres to this rule as well, so it's probably the best approach to stay aligned. If it ever becomes an issue for you, we'll find a solution.

Like pay for psychological counseling? ;)

BudsieBuds replied:
There's some awesome and free breathing exercises online. We believe in you! 👀

BudsieBuds commented on app/Config/Format.php:
```diff
@@ -25,8 +25,8 @@ class Format extends BaseConfig
         'application/json',
-        'application/xml', // machine-readable XML
-        'text/xml', // human-readable XML
+        'application/xml', // Machine-readable XML
```
I hadn’t realized this was part of the framework code. Keeping framework files unmodified is probably best.

objecttothis replied:
> I hadn’t realized this was part of the framework code. Keeping framework files unmodified is probably best.

Except for the custom validation rule class, everything in app/Config/ is from the framework.  On your dev box, you can see what is coming from the framework by drilling down into the vendor folder.

BudsieBuds replied:
I'll look into the framework files.

BudsieBuds commented on app/Config/Validation/OSPOSRules.php:
```diff
@@ -1,15 +1,16 @@
+ * @property Employee employee
+ * @property IncomingRequest request
+ */
+
```
Agreed. I'll update this.

BudsieBuds commented:
> OK, I had time to get through 27 of the files. I'll do more probably Monday, but look at the comments I made because IMO, we need to make a customization to the linter

Thanks for helping review this!

> - Enforce each brace `{` `}` getting it's own line always (see my comments about it.

As mentioned in my review replies, there are rules in the PSR-12 standard enforcing braces on the same line for control structures and on a new line for methods and functions.

> - No space between constructs (if, foreach, elseif, etc) and the opening parenthesis `(`

Same here. This is defined in the PSR-12 coding style.

> - No space between // and the comment.

I tried to add or keep the space between `//` and the comment. This should be consistent in the PR.

objecttothis commented on app/Config/Filters.php:
```diff
@@ -70,7 +70,7 @@ class Filters extends BaseFilters
         'before' => [
             'honeypot',
-            //'csrf' => ['except' => 'login'],    //TODO: Temporarily disable CSRF until we get everything sorted.
+            // 'csrf' => ['except' => 'login'],    // TODO: Temporarily disable CSRF until we get everything sorted
```
Fine.  I don't feel strongly enough about it.  I'll make a comment if I see any without a space... I think I did see a few comments without spaces.

BudsieBuds commented on app/Config/Pager.php:
```diff
@@ -43,19 +43,19 @@ class Pager extends BaseConfig
      */
     public $config = [
-        'full_tag_open' => "<ul class='pagination pagination-sm'>",
+        'full_tag_open' => '<ul class="pagination pagination-sm">',
```
Found that this is code added to the framework file. So it's not CI code and we can keep it.

objecttothis commented on app/Config/Mimes.php:
```diff
@@ -494,8 +494,7 @@ public static function guessTypeFromExtension(string $extension): array|string|n
 
-        if (!array_key_exists($extension, static::$mimes))
-        {
+        if (! array_key_exists($extension, static::$mimes)) {
```
That space though.... I couldn't find this in the PSR-12 documentation other than it says that code should adhere to PSR-1 and PSR-1 has an example with a space between the not operator and it's paired statement.

objecttothis replied:
OK, well let's just match the framework file to reduce workload with merging on update.

objecttothis reviewed (commented):
A few more comments. Sorry this is so piecemeal.  I don't have very much time for a full review until Thursday.

BudsieBuds commented on app/Config/Exceptions.php:
```diff
@@ -80,7 +80,7 @@ class Exceptions extends BaseConfig
     public string $deprecationLogLevel = LogLevel::WARNING;
 
-    /**
+    /*
```
Yes, I added it. But going over the CI framework files, found this to be missing there. So it should probably be fixed there and merged here. See here: https://github.com/codeigniter4/CodeIgniter4/blob/v4.6.0/app/Config/Exceptions.php

BudsieBuds commented on app/Config/Autoload.php:
```diff
@@ -63,112 +63,112 @@ class Autoload extends AutoloadConfig
-        'Employees' => '/App/Controllers/Employees.php',
-        'Expenses' => '/App/Controllers/Expenses.php',
+        // Controllers
+        'Attributes'          => '/App/Controllers/Attributes.php',
```
This is OSPOS code, so it shouldn't matter right? The CI framework code uses this alignment. I'm adjusting our code to have this alignment everywhere. Hopefully committing today.

objecttothis replied:
Yeah. Historically we didn't have the spacing in there because the tabs would cause some IDEs to display them incorrectly aligned. With 4 spaces I don't think that's an issue. I like it myself and if it's not going to cause merge or alignment issues then I'm fine with it.

objecttothis commented:
In the meantime, it looks like there are several files that need to have conflicts resolved before you will be able to merge.

objecttothis commented on writable/.htaccess:
```diff
@@ -1,6 +1,6 @@
 <IfModule authz_core_module>
-    Require all denied
+	Require all denied
```
Same issue here with tabs being added instead of 4 spaces

BudsieBuds replied:
Same as above, it's a framework file.

objecttothis commented on app/Models/Expense.php:
```diff
@@ -137,65 +134,55 @@ public function search(string $search, array $filters, ?int $rows = 0, ?int $lim
-            $builder->orLike('expenses.amount', $search);
-            $builder->orLike('expense_categories.category_name', $search);
-            $builder->orLike('CONCAT(employees.first_name, " ", employees.last_name)', $search);
+        $builder->like('employees.first_name', $search);
```
I know this isn't a part of a structure like if so technically isn't indented however it was indented to make it clear in the builder that this was all part of the same group.  IMO it should stay unless it causes the linter to fail during build.

BudsieBuds replied:
I agree, fixed.

BudsieBuds commented on app/.htaccess:
```diff
@@ -1,6 +1,6 @@
 <IfModule authz_core_module>
-    Require all denied
+	Require all denied
```
I fully agree. However, same as some conversations above, these changes where ported back from the framework files.

BudsieBuds commented on app/Language/ar-EG/Attributes.php:
```diff
@@ -1,32 +1,33 @@
-    "confirm_delete" => "هل أنت متأكد من أنك تريد حذف الميزات المحددة ؟",
-    "confirm_restore" => "هل أنت متأكد من أنك تريد استعادة السمة (السمات) المحددة؟",
-    "definition_cannot_be_deleted" => "لا يمكن حذف السمات المحددة",
+    "attribute_value_invalid_chars"    => "الميزات لا يمكن أن تحتوي على ':' أو'|'",
```
Yeah, I wondered this too. Do you know @jekkos?

BudsieBuds committed:
Coding style updates for PSR-12 compliacne & improved readability
- refactor code formatting for better PSR-12 standards
- unify coding conventions across the codebase for more consistent styling
- `else if` to `elseif`
- `boolean` to `bool`
- added missing framework files
- other minor updates

BudsieBuds committed:
Align arrays for improved readability
- reformatted arrays for improved readability

BudsieBuds committed:
Adjusted language files
- for consistency, update single quotation marks to double quotation marks
- adjusted spacing for readability and fixed some spacing
- add line break to match ci4 framework language file example

BudsieBuds commented on app/Controllers/Reports.php:
```diff
@@ -1360,32 +1323,30 @@ public function specific_customers(string $start_date, string $end_date, string
                     to_currency($drow['cost']),
                     to_currency($drow['profit']),
-                    ($drow['discount_type'] == PERCENT)? $drow['discount'].'%':to_currency($drow['discount'])
+                    ($drow['discount_type'] == PERCENT) ? $drow['discount'] . '%' : to_currency($drow['discount'])
```
Since the line is pretty short, I'll keep it like this for now. The trailing comma isn't mandated. It can be added for minimizing diffs when lines are added. But I don't think that's necessary for now.

BudsieBuds commented on app/Controllers/Reports.php:
```diff
@@ -1537,12 +1491,11 @@ public function specific_discount_input(): void
         }
         $data['specific_input_data'] = $discounts;
-        $data['discount_type_options'] = ['0' => lang('Reports.discount_percent'), '1'=> lang('Reports.discount_fixed')];
+        $data['discount_type_options'] = ['0' => lang('Reports.discount_percent'), '1' => lang('Reports.discount_fixed')];
```
Same as comment above, probably not long enough to split up.

BudsieBuds commented on app/Controllers/Reports.php:
```diff
@@ -1853,88 +1794,82 @@ public function detailed_sales(string $start_date, string $end_date, string $sal
-                        ($drow['discount_type'] == PERCENT) ? $drow['discount'].'%' : to_currency($drow['discount'])
-                    ],
+                    array_merge(
+                        [
```
I agree, fixed.

BudsieBuds committed:
Fixes

BudsieBuds commented:
@objecttothis I’ve rebased this PR to make it mergeable again. I think we should try to merge it soon. While it might not be perfectly consistent or fully PSR-12 compliant, it’s a step in the right direction.

objecttothis commented:
> @objecttothis I’ve rebased this PR to make it mergeable again. I think we should try to merge it soon. While it might not be perfectly consistent or fully PSR-12 compliant, it’s a step in the right direction.

Go for it!

BudsieBuds committed:
Merge branch 'master' into psr-12

BudsieBuds commented:
@objecttothis Had to resolve some conflicts. If you can review this, I can merge. 👍

jekkos commented:
@objecttothis Regarding the PSR compliance work - looking at the current state of the codebase, there are quite a lot of open PRs with significant changes:

- **Bootstrap 5** (#4445) - Draft, major UI overhaul
- **Plugin system** (#4407) - Open, architectural changes
- **Person attributes** (#4521) - Open, attribute system changes
- ~25 other open PRs

Trying to do a full PSR conversion at this point might create massive merge conflicts across many of these pending changes. 

Perhaps we should plan the PSR conversion for after some of the bigger changes like Bootstrap 5 and the plugin system are merged? That way the conversion can be done once without needing to reconcile conflicts across multiple branches.

What do you think about the timing here?
